ydb/library/yql/minikql/comp_nodes/mkql_wide_top_sort.cpp

```diff
@@ -393,146 +369,115 @@ using TBase = TComputationValue<TSpillingSupportState<Sort, HasCount>>;
             return;
         }
-        throw yexception() << "Spilling doesn't support TopSort.";
-    }
-
-    virtual EFetchResult DoCalculate(TComputationContext& ctx, NUdf::TUnboxedValue*const* output) {
-        while (true) {
-            switch(GetMode()) {
-                case EOperatingMode::InMemory: {
-                    auto r = DoCalculateInMemory(ctx, output);
-                    if (GetMode() == TSpillingSupportState::EOperatingMode::InMemory) {
-                        return r;
-                    }
-                    break;
-                }
-                case EOperatingMode::Spilling: {
-                    DoCalculateWithSpilling(ctx);
-                    if (GetMode() == EOperatingMode::Spilling) {
-                        return EFetchResult::Yield;
-                    }
-                    break;
-                }
-                case EOperatingMode::ProcessSpilled: {
-                    return ProcessSpilledData(output);
-                }
 
-            }
+        Storage.resize(GetStorageSize() * Indexes.size());
+        Free.resize(GetStorageSize(), nullptr);
+        if (Count) {
+            Full.reserve(GetStorageSize());
+            auto ptr = Storage.data();
+            std::generate(Free.begin(), Free.end(), [&ptr, this]() {
+                const auto p = ptr;
+                ptr += Indexes.size();
+                return p;
+            });
+            ResetFields();
+        } else {
+            InputStatus = EFetchResult::Finish;
         }
-        Y_UNREACHABLE();
     }
 
-private:
-
-    EFetchResult DoCalculateInMemory(TComputationContext& ctx, NUdf::TUnboxedValue*const* output) {
-        while (EFetchResult::Finish != InputStatus) {
-            switch (InputStatus = Flow->FetchValues(ctx, GetFields())) {
-                case EFetchResult::One:
-                    if (Put()) {
-                        if (ctx.SpillerFactory && !HasMemoryForProcessing()) {
-                            SwitchMode(EOperatingMode::Spilling, ctx);
-                            return EFetchResult::Yield;
-                        }
-                    }
-                    continue;
-                case EFetchResult::Yield:
-                    return EFetchResult::Yield;
-                case EFetchResult::Finish:
-                {
-                    if (!SpilledStates.empty()) {
-                        SwitchMode(EOperatingMode::Spilling, ctx);
-                        return EFetchResult::Yield;
+    bool IsReadyToContinue() {
+        switch (GetMode()) {
+            case EOperatingMode::InMemory:
+               return true;
+            case EOperatingMode::Spilling:
+            {
+                if (!SpillState()) {
+                    return false;
+                }
+                ResetFields();
+                auto nextMode = (IsReadFromChannelFinished() ? EOperatingMode::ProcessSpilled : EOperatingMode::InMemory);
+                SwitchMode(nextMode);
```

Comment:
As far as I remember log was intended here after switching the modes, is it missing?

Reply:
Oh, I see
Log conflicted here :(